### PR TITLE
Patcherex2 init

### PR DESCRIPTION
**Pull request by Ninja3047**
* Supports emitting thunks for our patch compiler
 * Supports rewriting addresses for our patch compiler
 * Update dependency to our keystone fork which fixes a bug with data directives and relative jumps

### CODE REVIEW

File: src/patcherex2/components/utils/utils.py [hunk 1]
```diff
@@ -11,7 +12,13 @@ def __init__(self, p, binary_path):
...
 
     def insert_trampoline_code(
-        self, addr, instrs, force_insert=False, detour_pos=-1, symbols=None
+        self,
+        addr,
+        instrs,
+        force_insert=False,
+        detour_pos=-1,
+        symbols=None,
+        base_reg=None,
```
Comment (2over12) [inline src/patcherex2/components/utils/utils.py]: fixing a base reg and this behavior here feels weird

File: src/patcherex2/components/utils/utils.py [hunk 2]
```diff
@@ -59,8 +78,65 @@ def insert_trampoline_code(
...
+                if re.match(r"\.(long) [0-9]+", line):
+                    _, goto_addr = line.split(" ")
+                    goto_addr = int(goto_addr)
+                    # only rewrite goto addresses in between the start of the moved instructions
+                    # to the end of the moved instructions
+                    if (
+                        goto_addr - addr >= 0
+                        and goto_addr - addr <= self.p.target.JMP_SIZE
+                    ):
+                        instrs_len = (
```
Comment (2over12) [inline src/patcherex2/components/utils/utils.py]: ive been staring at this line and i cant really figure what it's trying to compute. At first i thought it was the user instruction length so we could find [thunk][user code][old code][jump back]. If im thinking about this right goto_addr-addr should be the offset we want into the old code. So we want  old_code offset = trampoline_size - (jump size + moved instruction length). Then old_code_offset + mem_addr + (goto_addr - addr). Where is this 4 coming from im a bit lost here

File: src/patcherex2/targets/elf_ppc_linux.py [hunk 3]
```diff
@@ -74,3 +74,36 @@ def get_utils(self, utils):
...
             return Utils(self.p, self.binary_path)
         raise NotImplementedError()
+
+    # Emits a thunk which calculates the base address
+    # by subtracting the patch insertion address from
+    # the current pc and stores it in the given register
+    @staticmethod
+    def emit_thunk(base_reg, insert_addr, is_thumb=False):
+        assert "%r" in base_reg
+        scratch_reg1 = "%r4" if base_reg == "%r3" else "%r3"
```
Comment (2over12) [inline src/patcherex2/targets/elf_ppc_linux.py:83]: does keystone enable register prefixes on ppc?

Comment (Ninja3047) [reply src/patcherex2/targets/elf_ppc_linux.py:83]: parsing-wise, it supports both transparently
https://github.com/keystone-engine/keystone/blob/master/llvm/lib/Target/PowerPC/AsmParser/PPCAsmParser.cpp#L1517

mixing in theory should be fine but i just have it assert for now, probably will end up removing it once i know for sure that it'll work

File: src/patcherex2/components/utils/utils.py (hunk 2 above)
Comment (Ninja3047) [reply src/patcherex2/components/utils/utils.py]: simplified the logic a bit in the new version
the previous logic was basically trying to undo the math they did [here](https://github.com/trail-of-forks/Patcherex2/blob/2e6f9e106d52c32c691caae95f84e0b30d2b1b3b/src/patcherex2/components/utils/utils.py#L47) to calculate the instruction length
but i rewrote it so it doesn't do that anymore

**Review by 2over12 — APPROVED** [review]

**Comment by Ninja3047** [comment]
we now can use upstream patcherex2 with no modifications

**Pull request closed without merging**